src/rules/valid-context-access.ts

```diff
@@ -0,0 +1,223 @@
+      belongingFunction:
+        | TSESTree.FunctionDeclaration
+        | TSESTree.VariableDeclaration
+        | TSESTree.ArrowFunctionExpression
+      node: TSESTree.Node
+    }[] = []
+
+    /** Return true if the given node is later than the await expression. */
+    function isAfterAwait(node: TSESTree.CallExpression) {
+      for (const awaitExpression of awaitExpressions) {
+        const { belongingFunction, node: awaitNode } = awaitExpression
+        if (isInsideOf(node, belongingFunction)) {
+          continue
+        }
+        return awaitNode.range[0] <= node.range[0]
+      }
+      return false
+    }
+
+    /** Return true if node is inside of task function */
+    function isInsideTaskReference(node: TSESTree.CallExpression) {
+      for (const taskReference of taskReferences) {
+        if (isInsideOf(taskReference.node, node)) {
+          return true
+        }
+      }
+      return false
+    }
+
+    /** Let's lint! */
+    function doLint(
+      visitedCallExpressions: TSESTree.CallExpression[],
+      contextCallExpression: TSESTree.CallExpression,
+      currentNode: TSESTree.CallExpression,
+    ) {
+      // Report if context function is called outside of <script> block.
+      if (!isInsideOfSvelteScriptElement(currentNode)) {
+        report(contextCallExpression)
+        return
+      }
+
+      if (isAfterAwait(currentNode)) {
+        report(contextCallExpression)
+        return
+      }
+
+      if (isInsideTaskReference(currentNode)) {
+        report(contextCallExpression)
+        return
+      }
+
+      if (isInsideOfPromiseThenOrCatch(currentNode)) {
+        report(contextCallExpression)
+        return
+      }
+
+      let { parent } = currentNode
+      while (parent) {
+        parent = parent.parent
+        if (
+          parent?.type === "VariableDeclaration" ||
+          parent?.type === "FunctionDeclaration"
+        ) {
+          const references =
+            parent.type === "VariableDeclaration"
+              ? getReferences(parent.declarations[0].id)
+              : parent.id
+              ? getReferences(parent.id)
+              : []
+
+          for (const reference of references) {
+            if (reference.identifier?.parent?.type === "CallExpression") {
+              if (
+                !visitedCallExpressions.includes(reference.identifier.parent)
+              ) {
+                visitedCallExpressions.push(reference.identifier.parent)
+                doLint(
+                  visitedCallExpressions,
+                  contextCallExpression,
+                  reference.identifier?.parent,
+                )
+              }
+            }
+          }
+        } else if (parent?.type === "ExpressionStatement") {
+          if (parent.expression.type !== "CallExpression") {
+            report(contextCallExpression)
+          } else if (lifeCycleReferences.includes(parent.expression)) {
+            report(contextCallExpression)
+          }
+        }
```

Comment:
I think it's safer to check the VariableDeclarator. Also, I think we should check if the right operand is a function expression.
Otherwise, the following cases will result in false positives.

```svelte
<script>
  import { getContext, onMount } from "svelte"

  const foo = getContext("foo")

  onMount(() => {
    foo()
  })
</script>

```

```suggestion
        if (
          (parent?.type === "VariableDeclarator" &&
            parent.init &&
            (parent.init.type === "FunctionExpression" ||
              parent.init.type === "ArrowFunctionExpression") &&
            isInsideOf(parent.init, currentNode)) ||
          parent?.type === "FunctionDeclaration"
        ) {
          const references = parent.id ? getReferences(parent.id) : []

          for (const reference of references) {
            if (reference.identifier?.parent?.type === "CallExpression") {
              if (
                !visitedCallExpressions.includes(reference.identifier.parent)
              ) {
                visitedCallExpressions.push(reference.identifier.parent)
                doLint(
                  visitedCallExpressions,
                  contextCallExpression,
                  reference.identifier?.parent,
                )
              }
            }
          }
        } else if (parent?.type === "ExpressionStatement") {
          if (parent.expression.type !== "CallExpression") {
            report(contextCallExpression)
          } else if (lifeCycleReferences.includes(parent.expression)) {
            report(contextCallExpression)
          }
        }
        parent = parent.parent
```